prog/app.go

```diff
@@ -40,14 +42,81 @@ func appMain() {
+		controlRouterType = flag.String("control.router", "local", "Control router to use (local or sqs)")
+		pipeRouterType    = flag.String("pipe.router", "local", "Pipe router to use (local)")
+		userIDHeader      = flag.String("userid.header", "", "HTTP header to use as userid")
+
+		awsDynamoDB     = flag.String("aws.dynamodb", "", "URL of DynamoDB instance")
+		awsCreateTables = flag.Bool("aws.create.tables", false, "Create the tables in DynamoDB")
+		awsSQS          = flag.String("aws.sqs", "", "URL of SQS instance")
+		awsRegion       = flag.String("aws.region", "", "AWS Region")
+		awsID           = flag.String("aws.id", "", "AWS Account ID")
+		awsSecret       = flag.String("aws.secret", "", "AWS Account Secret")
+		awsToken        = flag.String("aws.token", "", "AWS Account Token")
+
+		consulPrefix = flag.String("consul.prefix", "", "Prefix for keys in consul")
+		consulAddr   = flag.String("consul.addr", "", "Address of consul instance")
+		consulInf    = flag.String("consul.inf", "", "The interface who's address I should advertise myself under in consul")
 	)
 	flag.Parse()
 
 	setLogLevel(*logLevel)
 	setLogFormatter(*logPrefix)
 
-	defer log.Info("app exiting")
+	// Do we need a user IDer?
+	var userIDer = multitenant.NoopUserIDer
+	if *userIDHeader != "" {
+		userIDer = multitenant.UserIDHeader(*userIDHeader)
+	}
+
+	// Create a collector
+	var collector app.Collector
+	if *collectorType == "local" {
+		collector = app.NewCollector(*window)
+	} else if *collectorType == "dynamodb" {
+		creds := credentials.NewStaticCredentials(*awsID, *awsSecret, *awsToken)
+		dynamoCollector := multitenant.NewDynamoDBCollector(*awsDynamoDB, *awsRegion, creds, userIDer)
+		collector = dynamoCollector
+		if *awsCreateTables {
+			if err := dynamoCollector.CreateTables(); err != nil {
+				log.Fatalf("Error createing DynamoDB tables: %v", err)
+			}
+		}
+	} else {
+		log.Fatalf("Invalid collector '%s'", *collectorType)
+		return
+	}
 
+	// Create a control router
+	var controlRouter app.ControlRouter
+	if *controlRouterType == "local" {
+		controlRouter = app.NewLocalControlRouter()
+	} else if *controlRouterType == "sqs" {
+		creds := credentials.NewStaticCredentials(*awsID, *awsSecret, *awsToken)
+		controlRouter = multitenant.NewSQSControlRouter(*awsSQS, *awsRegion, creds, userIDer)
+	} else {
+		log.Fatalf("Invalid control router '%s'", *controlRouterType)
+		return
+	}
+
+	// Create a pipe router
+	var pipeRouter app.PipeRouter
+	if *pipeRouterType == "local" {
+		pipeRouter = app.NewLocalPipeRouter()
+	} else if *pipeRouterType == "consul" {
+		var err error
+		pipeRouter, err = multitenant.NewConsulPipeRouter(*consulAddr, *consulPrefix, *consulInf, userIDer)
+		if err != nil {
+			log.Fatalf("Error createing consul pipe router: %v", err)
+		}
+	} else {
+		log.Fatalf("Invalid pipe router '%s'", *pipeRouterType)
+		return
+	}
+
+	defer log.Info("app exiting")
```

Comment:
Not putting them inline ends up with a bunch of factory functions (eugh) with lots of arguments.

This function is nothing compares to the probe main...

When we add help, I expect I'll factor all the args out into a struct, which would make the factory functions slightly less ugly.